rust/src/sdp/parser.rs

```diff
@@ -0,0 +1,675 @@
+    pub addrtype: String,
+    pub connection_address: IpAddr,
+    pub ttl: Option<u8>,
+    pub number_of_addresses: Option<u8>,
+}
+
+#[derive(Debug)]
+pub struct MediaDescription {
+    pub media: String,
+    pub port: u16,
+    pub number_of_ports: Option<u16>,
+    pub proto: String,
+    pub fmt: Vec<String>,
+    pub session_info: Option<String>,
+    pub connection_data: Option<ConnectionData>,
+    pub bandwidths: Option<Vec<String>>,
+    pub encryption_key: Option<String>,
+    pub attributes: Option<Vec<String>>,
+}
+
+// token-char = %x21 / %x23-27 / %x2A-2B / %x2D-2E / %x30-39 / %x41-5A / %x5E-7E
+#[inline]
+fn is_token_char(b: u8) -> bool {
+    matches!(b, 0x21 | 0x2A | 0x2B | 0x2D | 0x2E)
+        || (0x23..=0x27).contains(&b)
+        || (0x30..=0x39).contains(&b)
+        || (0x41..=0x5a).contains(&b)
+        || (0x5e..=0x7e).contains(&b)
+}
+
+#[inline]
+fn is_request_uri_char(b: u8) -> bool {
+    is_alphanumeric(b) || is_token_char(b) || b"~#@:;=?+&$,/".contains(&b)
+}
+
+#[inline]
+fn is_line_ending(b: u8) -> bool {
+    b == b'\r' || b == b'\n'
+}
+
+#[inline]
+fn is_ipaddr(b: u8) -> bool {
+    b.is_ascii_hexdigit() || b".:".contains(&b)
+}
+
+#[inline]
+fn is_session_name_char(b: u8) -> bool {
+    is_alphanumeric(b) || is_space(b)
+}
+
+#[inline]
+fn is_time_char(b: u8) -> bool {
+    is_digit(b) || b"dhms-".contains(&b)
+}
+
+fn parse_num(i: &[u8]) -> IResult<&[u8], u8> {
+    let (i, first_digit) = peek(be_u8)(i)?;
+    if first_digit == 0 {
+        return Err(Err::Error(make_error(i, ErrorKind::HexDigit)));
+    }
+    let (i, num) = take_u8(i)?;
+    Ok((i, num))
+}
+
+// SDP Message format (fields marked with * are optional):
+// https://www.rfc-editor.org/rfc/rfc4566#page-9
+//
+//       Session description
+//         v=  (protocol version)
+//         o=  (originator and session identifier)
+//         s=  (session name)
+//         i=* (session information)
+//         u=* (URI of description)
+//         e=* (email address)
+//         p=* (phone number)
+//         c=* (connection information -- not required if included in
+//              all media)
+//         b=* (zero or more bandwidth information lines)
+//         One or more time descriptions ("t=" and "r=" lines; see below)
+//         z=* (time zone adjustments)
+//         k=* (encryption key)
+//         a=* (zero or more session attribute lines)
+//         Zero or more media descriptions
+//
+//      Time description
+//         t=  (time the session is active)
+//         r=* (zero or more repeat times)
+//
+//      Media description, if present
+//         m=  (media name and transport address)
+//         i=* (media title)
+//         c=* (connection information -- optional if included at
+//              session level)
+//         b=* (zero or more bandwidth information lines)
+//         k=* (encryption key)
+//         a=* (zero or more media attribute lines)
+
+pub fn sdp_parse_message(i: &[u8]) -> IResult<&[u8], SdpMessage> {
```

Comment:
this parser requires the order to be like this, right? Can traffic have a different order?

Reply:
Right. The order must be followed exactly and cannot be changed.